File: frontend/types/dashboard/slotViz.ts
export type SlotVizIcons = 'attestation' | 'head_attestation' | 'source_attestation' | 'target_attestation' | 'proposal' | 'slashing' | 'sync'

export type SlotVizValidatorDuties = {
  type: 'proposal' | 'slashing' | 'attestation' | 'sync'
  pendingCount?: number,
  successCount?: number,
  successEarning?: string, // wei
  failedCount?: number,
  failedEarnings?: string, // wei
  validator?: number // in case it's a proposal
}

export type SlotVizSlot = {
  id: number
  state: 'orphaned' | 'missed' | 'proposed' | 'scheduled'
  duties?: SlotVizValidatorDuties[] // in case of a dashboard slotviz, the duties of the dashboard's validators
}

export type SlotVizEpoch = {
  id: number
  state: 'head' | 'finalized' | 'scheduled'
  slots: SlotVizSlot[]
}

export type SlotVizData = {
  currentSlot: number
  epochs: SlotVizEpoch[]
}

File: backend/api/structs/validator_dashboard/slot_viz.go
package api

import (
	"github.com/shopspring/decimal"
)

type VDBSlotViz struct {
	CurrentSlot uint64            `json:"current_slot"`
	Epochs      []VDBSlotVizEpoch `json:"epochs"`
}

type VDBSlotVizEpoch struct {
	Number uint64           `json:"number"`
	State  string           `json:"state"` // head, finalized, scheduled
	Slots  []VDBSlotVizSlot `json:"slots"`
}

type VDBSlotVizSlot struct {
	Number uint64           `json:"number"`
	Status string           `json:"status"` // proposed, missed, scheduled, orphaned
	Duties []VDBSlotVizDuty `json:"duties"`
}

type VDBSlotVizDuty struct {
	Type            string          `json:"type"` // proposal, attestation, sync, slashing
	PendingCount    uint64          `json:"pending_count"`
	SuccessCount    uint64          `json:"success_count"`
	SuccessEarnings decimal.Decimal `json:"success_earnings"`
	FailedCount     uint64          `json:"failed_count"`
	FailedEarnings  decimal.Decimal `json:"failed_earnings"`
	Validator       uint64          `json:"validator"`
	Block           uint64          `json:"block"`
}
